feat: 위젯 통해서 앱 열기 (#470)

hyun99999 opened the pull request:
## 🌴 PR 요약

🌱 작업한 브랜치
- #470
- #489

🌱 작업한 내용
- qr 스캔 위젯
  - qr 를 스캔할 수 있는 화면이 등장합니다.
- 내 명함 위젯
  - 내 명함을 바로 공유할 수 있는 바텀시트가 등장합니다.
- 추가적으로 해당 브랜치에서 #487 도 진행하였습니다.

## 🚨 참고사항
IntentHandler 의 defaultMyCard 메서드가 이제 위젯을 추가할때 호출되는데 이때 서버통신을 통해서 대표명함을 찾고, 이를 기본값으로 설정해주도록 구현하려고 했습니다.
즉, 서버통신을 통해서 결과데이터를 가져오면 그때 리턴하려고했는데 completion handler 내부에서 return 할 수 없기 때문에 고민입니다.. 방법이 있을까요?

## 📸 스크린샷
|기능|스크린샷|
|:--:|:--:|
|내 명함 위젯|<img src="https://user-images.githubusercontent.com/69136340/236623995-e70e5d09-8428-49e8-a3f4-b89b8450bdb4.png" width ="250">|
|위젯에서 앱 열기|<img src="https://user-images.githubusercontent.com/69136340/236624070-a1529dba-ef92-446d-8680-e67d96de88da.mov" width ="250">|

## 📮 관련 이슈
- Resolved: #470
- - Resolved: #489

hyun99999 committed:
feat: 매개면수 수정 (#470)

- cardUUID 로 수정

hyun99999 committed:
fix: 매개변수 수정 및 widgetURL 수정 (#470)

- cardID 에서 cardUUID 로 수정.

hyun99999 committed:
feat: 내 명함/qr코드 위젯을 통해 앱에서 기능 구현 (#470)

hyun99999 committed:
feat: 명함 조회 후 명함공유 바텀시트 등장 (#470)

dlwns33 committed:
feat: 테스트플라이트용 위젯 수정

hyun99999 committed:
Merge branch 'feature/justincase' of https://github.com/dlwns33/NADA-iOS-ForRelease into feature/#470

hyun99999 committed:
chore: pod update (#470)

hyun99999 committed:
fix: 온보딩, 액세스 토큰 UserDefaults key 변경 (#489)

hyun99999 committed:
feat: 앱 런치 시 위젯 접근 감지 (#470)

hyun99999 committed:
add: 위젯 관련 UserDefaults key 추가 (#470)

hyun99999 committed:
feat: 홈에서 위젯을 통해 QR 스캔 화면전환 (#470)

hyun99999 committed:
feat: 홈에서 위젯을 통해 명함 공유 화면전환 (#470)

hyun99999 committed:
feat: UserDefaults 를 활용해서 업데이트 판단 후에 위젯통해서 접근 구현 (#470)

hyun99999 committed:
delete: 기본동작에서 서버통신 삭제 (#470)

hyun99999 committed:
refactor: 내 명함 목록 선택 구현 리펙토링 (#470)

- result.data 를 옵셔널 바인딩하여 바로 사용.

hyun99999 committed:
feat: 위젯 기본값 설정 구현 (#470)

hyun99999 committed:
Merge branch 'develop' of https://github.com/TeamNADA/NADA-iOS-ForRelease into feature/#470

dlwns33 commented on IntentsExtension/IntentHandler.swift:
```diff
@@ -17,53 +17,51 @@ class IntentHandler: INExtension {
+        cardListFetchWithAPI { [weak self] result in
+            switch result {
+            case .success(let result):
+                if let result {
+                    self?.cardItems = result.data
+                    myCard = MyCard(identifier: self?.cardItems?[0].cardUUID ?? "", display: self?.cardItems?[0].cardName ?? "")
+                    myCard?.userName = self?.cardItems?[0].userName
+                    myCard?.cardImage = self?.cardItems?[0].cardImage
```
교수님! 참고사항에 적힌 내용이 여기에서 return을 하려고 했는데 못한다는 얘기일까요?

hyun99999 replied:
그쳐.. 서버통신이 비동기라서 리턴이랑은 별개로 알아서 되는거라 nil 값이 리턴됩디다

hyun99999 replied:
@dlwns33 
dispatch group 으로 해결했심더!

- 서버통신의 결과를 동기적으로 사용하기 위해서 DispatchGroup 을 사용.
- 비동기 통신이기 때문에 group 의 enter,leave 사용.
- 60초까지만 기다리기로 함.

hyun99999 replied:
- 액세스 토큰인 있는 경우에 위젯의 기본값 설정
<img src="https://user-images.githubusercontent.com/69136340/236741933-57bff414-46f1-416b-8a83-8c0afd2e7b6a.mov" width ="200">

hyun99999 replied:
- 액세스 토큰이 있는 경우에 위젯의 기본값 설정

<img src="https://user-images.githubusercontent.com/69136340/236742586-d21301c8-472f-4750-9146-102e75e0b909.mp4" width ="200">

hyun99999 replied:
- 액세스 토큰이 없어서 명함을 선택할 수 없는 경우
<img src="https://user-images.githubusercontent.com/69136340/236743158-6e16b114-5444-4c44-a756-46217f823dba.mov" width ="200">

hyun99999 committed:
feat: 위젯 기본값 설정 (#470)

- 서버통신의 결과를 동기적으로 사용하기 위해서 DispatchGroup 을 사용.
- 비동기 통신이기 때문에 group 의 enter,leave 사용.
- 60초까지만 기다리기로 함.

hyun99999 committed:
feat: 내 명함 선택 목록 구성 실패시에 nil 반환 (#470)

hyun99999 committed:
feat: 서버통신 성공 후 status 200 아닌 에러 잡기 (#470)
